scenario/state.py

```diff
@@ -1772,17 +1783,17 @@ def deferred(self, handler: Callable, event_id: int = 1) -> DeferredEvent:
         owner_name, handler_name = match.groups()[0].split(".")[-2:]
         handle_path = f"{owner_name}/on/{self.name}[{event_id}]"
 
+        # Many events have no snapshot data: install, start, stop, remove, config-changed,
+        # upgrade-charm, pre-series-upgrade, post-series-upgrade, leader-elected,
+        # leader-settings-changed, collect-metrics
         snapshot_data = {}
 
         # fixme: at this stage we can't determine if the event is a builtin one or not; if it is
         #  not, then the coming checks are meaningless: the custom event could be named like a
         #  relation event but not *be* one.
         if self._is_workload_event:
-            # this is a WorkloadEvent. The snapshot:
-            container = cast(Container, self.container)
-            snapshot_data = {
-                "container_name": container.name,
-            }
+            assert self.container is not None
```

Comment:
When the only goal of an assert is to type guard, I personally try to use `cast`.
If you like asserts more, fine, but then I'd add a `# type guard` to explain what that's doing.
Otherwise as we know, asserts in runtime code are weird

Reply:
my 2c: I prefer `assert` in this case.

It's a testing library, if somehow this assert fails, then test fails and it's reasonably clear what has happened

Additionally a comment from @benhoyt: assert is fine because it's clear what it means in the context.

Reply:
Yeah, as Dima hinted, I prefer `assert` here, because it states the intent much more clearly than `cast()`. "assert something is not None" says what we mean, where "cast()" implies it's coercing to a different type.

We don't have to worry about assertions being optimised out, as this isn't a runtime check anyway -- it's a static type checking thing (or: type checkers aren't the ones that optimise out asserts).

Reply:
ok, I'm convinced, I rest my case.